storybookjs/storybook
# Release tooling: Defer version bumping to the publish step

**I will manually cherry-pick this back to `main` in follow-up PR**



## What I did

This PR changes the release flow to defer the actual version bump to the publish step instead of when preparing  the release PRs. This makes it possible to run CI on the release PRs. It will still calculate the version bump up front.

The `release:version` script now take two additional flags:

- `--deferred`: This will keep the same bumping logic, except instead of bumping the versions in all the packages it will simply set a `"deferredNextVersion"` property in `code/package.json` and exit.
- `--apply`: This will apply the version set at `"deferredNextVersion"`, by actually bumping all the package's versions as usual. It's invalid to use this flag with `--release-type` or `--pre-id`, since `--apply` will not do any increment calculation.

The release preparation workflows then uses `--deferred` to set the deferred version.
The publish workflow will look for the `"deferredNextVersion"` and if it is set, it will run `release:version` with `--apply` and commit the changes. It will not be set in cases where patch releases don't contain any releasable changes.

This deferred workflow doesn't make much sense to run locally, and it's not necessary to do so, but it can be if you want to.



## How to test

- `cd scripts`
- `yarn release:version --release-type preminor --pre-id test`
- See that `"deferredNextVersion"` is not set in `code/package.json`
- `yarn release:version --apply`
- See that all packages have been bumped and that `"deferredNextVersion"` is removed again



## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
